### PR TITLE
Remove JS preprocessor code from the web/ and src/ files

### LINKED ISSUE #7192: Remove all preprocessor directives from UMD modules
The preprocessor directives shall be remove from UMD modules (files in src/ and web/ folders). That is needed for migration of the modules to ES6 syntax. So far we have:

//#else
//#endif
//#expand ....
//#if ...
//#include ...

- [ ] e.g. Uglify might perform constant folding
- [ ] 'if'/'else'/'endif' shall be replaced with the corresponding JS statements
- [ ] 'expand' can be just plain JS without '__' suffix. 
- [ ] 'include'ed file must be replaced by modules or duplicated and inlined.

### PR DESCRIPTION
**Pull request by yurydelendik**
Fixes #7192

It introduces esprima-based processor and fully replaces our during bundling process. First commit contains just a preprocessor. Second replaces it in gulpfile and changes directives into new format (I did not re-indent large pieces of code).

```
//#if TEST
  jsCode1();
//#endif
//#if TEST
//jsCode2();
//#endif
```

was replaced by

```
if (typeof PDFJSDev === 'undefined' || PDFJSDev.test('TEST')) {
  jsCode1();
}
if (typeof PDFJSDev !== 'undefined' && PDFJSDev.test('TEST')) {
  jsCode2();
}
```

The preprocessor is just "smart" enough to fold these constructs above (and few more related one) -- it's not really-really smart though. See its tests at external/builder/fixtures2.

### CODE REVIEW

**Comment by yurydelendik** [comment]
TODOs:
- [ ] Test with extensions

File: extensions/firefox/content/PdfJsNetwork.jsm [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/* globals Components, Services, dump */
+
+// NOTE: Be careful what goes in this file, as it is also used from the context
+// of the addon. So using warn/error in here will break the addon.
```
Comment (Snuffleupagus) [inline extensions/firefox/content/PdfJsNetwork.jsm]: Nit: given that this file is now placed in `extensions/firefox/`, this comment no longer seem necessary.

File: extensions/firefox/content/PdfJsNetwork.jsm [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+
+'use strict';
+
+Components.utils.import('resource://gre/modules/Services.jsm');
+
+var EXPORTED_SYMBOLS = ['NetworkManager'];
+
+function log(aMsg) {
```
Comment (Snuffleupagus) [inline extensions/firefox/content/PdfJsNetwork.jsm]: Nit: do we still need this function for potential debugging purposes, or can it be removed now?

File: web/app.js [hunk 3]
```diff
@@ -1605,20 +1591,16 @@ function webViewerPageRendered(e) {
...
 
-//#if !PRODUCTION
-  if (true) {
-    return;
-  }
-//#endif
-//#if (FIREFOX || MOZCENTRAL)
-  if (pageView.textLayer && pageView.textLayer.textDivs &&
+  if (typeof PDFJSDev !== 'undefined' &&
+      !PDFJSDev.test('FIREFOX || MOZCENTRAL') &&
```
Comment (Snuffleupagus) [inline web/app.js]: Remove the `!`.

File: web/app.js [hunk 4]
```diff
@@ -489,8 +492,11 @@ var PDFViewerApplication = {
         PDFViewerApplication.progress(loaded / total);
       }
     });
+} else {
+    throw new Error('Not implemented: initPassiveLoading');
+}
   },
-//#endif
+
```
Comment (Snuffleupagus) [inline web/app.js]: Nit: unnecessary newline.

File: web/app.js [hunk 5]
```diff
@@ -1156,16 +1152,13 @@ var PDFViewerApplication = {
       this.pdfViewer.getPageView(i).beforePrint(printContainer);
     }
 
-//#if !PRODUCTION
-    if (true) {
-      return;
+
```
Comment (Snuffleupagus) [inline web/app.js]: Nit: unnecessary newline.

File: extensions/firefox/content/PdfJsNetwork.jsm (hunk 2 above)
Comment (yurydelendik) [inline extensions/firefox/content/PdfJsNetwork.jsm]: Let's keep it just in case. I hope this module will be refactored and this might be helpful.

File: src/display/api.js [hunk 6]
```diff
@@ -68,13 +68,39 @@ var isWorkerDisabled = false;
...
-//var fakeWorkerFilesLoader = null;
-//#endif
-//#endif
+var fakeWorkerFilesLoader = null;
+var useRequireEnsure = false;
+if (typeof PDFJSDev !== 'undefined' &&
+    PDFJSDev.test('GENERIC && !SINGLE_FILE')) {
+  // For GENERIC build we need add support of different fake file loaders
+  // for different  frameworks.
+  if (typeof window === 'undefined') {
```
Comment (Snuffleupagus) [inline src/display/api.js]: With this code being inlined here, do we still need to keep https://github.com/mozilla/pdf.js/blob/master/src/frameworks.js around?

Comment (yurydelendik) [inline src/display/api.js]: No, we don't. Removed.

File: .jshintignore [hunk 7]
```diff
@@ -9,6 +9,7 @@ external/jpgjs/
 external/jasmine/
 external/cmapscompress/
 external/importL10n/
+external/builder/fixtures2/
```
Comment (timvandermeij) [inline .jshintignore]: I'm not really liking the name `fixtures2` as it does not seem very descriptive. Can we change it to, for example, `fixtures_esprima` or something like that to make it more clear that it is about the Esprima preprocessor and not our own?

File: external/builder/preprocessor2.js [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+
+'use strict';
+
+var esprima = require('esprima');
+var escodegen = require('escodegen');
+var vm = require('vm');
+var fs = require('fs');
+var path = require('path');
+
+
```
Comment (timvandermeij) [inline external/builder/preprocessor2.js]: Nit: unnecessary newline

File: external/builder/preprocessor2.js [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+  return obj.type === 'Literal' && obj.value === value;
+}
+
+function isPDFJSPreprocessor(obj) {
+  return obj.type === 'Identifier' &&
+         obj.name === PDFJS_PREPROCESSOR_NAME;
+}
+
+function evalWithDefines(code, defines, loc) {
+    if (!code || !code.trim()) {
```
Comment (timvandermeij) [inline external/builder/preprocessor2.js]: The indentation is off in this function. Four spaces are used instead of two.

File: external/builder/preprocessor2.js [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+    }
+    throw new Error('Unsupported action');
+  } catch (e) {
+    throw new Error('Could not process ' + PDFJS_PREPROCESSOR_NAME + '.' +
+                    actionName + ' at ' + JSON.stringify(loc) + '\n' +
+                    e.name + ': ' + e.message);
+  }
+}
+
+function preprocessNode(ctx, node) {
```
Comment (timvandermeij) [inline external/builder/preprocessor2.js]: I wonder why we need this if we are directly returning the argument?

File: make.js [hunk 11]
```diff
@@ -104,6 +104,7 @@ var COMMON_WEB_FILES =
       ['web/viewer.html'],
     COMMON_FIREFOX_FILES_PREPROCESS =
       [FIREFOX_CONTENT_DIR + 'PdfStreamConverter.jsm',
+       FIREFOX_CONTENT_DIR + 'PdfJsNetwork.jsm',
```
Comment (timvandermeij) [inline make.js]: Why is the `J` captalized here? Other files (e.g. the two below this line) do not have that.

Comment (yurydelendik) [inline make.js]: Similar to other names "PdfJs.jsm" and "PdfJsTelemetry.jsm". We don't have consistency, do we?

File: src/display/font_loader.js [hunk 12]
```diff
@@ -101,13 +104,16 @@ FontLoader.prototype = {
...
-  addNativeFontFace: function fontLoader_addNativeFontFace(nativeFontFace) {
+  FontLoader.prototype.addNativeFontFace =
+      function fontLoader_addNativeFontFace(nativeFontFace) {
     this.nativeFontFaces.push(nativeFontFace);
     document.fonts.add(nativeFontFace);
-  },
+  };
 
-  bind: function fontLoaderBind(fonts, callback) {
+  FontLoader.prototype.bind = function fontLoaderBind(fonts, callback) {
```
Comment (timvandermeij) [inline src/display/font_loader.js]: What is the reason for this change?

Comment (yurydelendik) [inline src/display/font_loader.js]: The `FontLoader.prototype` object literal was closed above, so JavaScript if-statement could be used. ("if" cannot be used inside object literal)

**Comment by timvandermeij** [comment]
> I did not re-indent large pieces of code

What is the reason for that exactly? I think it would be good to have consistent indenting. We can ignore indentation changes during review with the `?w=1` flag, so there should be no concern of more complicated review.

**Comment by yurydelendik** [comment]
> > I did not re-indent large pieces of code

> What is the reason for that exactly?

I wanted to reduce amount of code affected by me, plus I'm sure we will change it again once starting migrating to ES modules.

File: make.js (hunk 11 above)
Comment (timvandermeij) [inline make.js]: I see. Never mind this comment in that case.

File: src/display/text_layer.js [hunk 13]
```diff
@@ -132,9 +132,8 @@ var renderTextLayer = (function renderTextLayerClosure() {
     }
 
     var canvas = document.createElement('canvas');
-//#if MOZCENTRAL || FIREFOX || GENERIC
-    canvas.mozOpaque = true;
-//#endif
+    canvas.mozOpaque = typeof PDFJSDev === 'undefined' ||
+                       PDFJSDev.test('FIREFOX || MOZCENTRAL || GENERIC');
```
Comment (Snuffleupagus) [inline src/display/text_layer.js]: Nit: this uses a slightly different form than the other `mozOpaque` occurrences, can we use the same one everywhere?

**Comment by yurydelendik** [comment]
Re-indented all affected by my code (probably except URL polyfill)

File: src/shared/util.js [hunk 14]
```diff
@@ -1679,7 +1682,7 @@ function loadJpegStream(id, imageUrl, objs) {
   img.src = imageUrl;
 }
 
-//#if !(MOZCENTRAL)
+if (typeof PDFJSDev === 'undefined' || !PDFJSDev.test('MOZCENTRAL')) {
 //// Polyfill from https://github.com/Polymer/URL
```
Comment (Snuffleupagus) [inline src/shared/util.js]: Nit: with the new proprocessor we shouldn't need double comments here, so `////` can be changed to `//`.

File: src/worker_loader.js [hunk 15]
```diff
@@ -15,16 +15,16 @@
...
-self.importScripts = (function (importScripts) {
-  return function() {
-    setTimeout(function () {}, 0);
-    return importScripts.apply(this, arguments);
-  };
-})(importScripts);
-//#endif
+if (typeof PDFJSDev === 'undefined' || !PDFJSDev.test('PRODUCTION')) {
+  //// Patch importScripts to work around a bug in WebKit and Chrome 48-.
+  //// See https://crbug.com/572225 and https://webkit.org/b/153317.
```
Comment (Snuffleupagus) [inline src/worker_loader.js]: Nit: with the new proprocessor we shouldn't need double comments here, so `////` can be changed to `//`.

Comment (yurydelendik) [inline src/worker_loader.js]: Fixed

**Comment by brendandahl** [comment]
r+ after rebase and tests

**Comment by yurydelendik** [comment]
Rebased

/botio test

File: web/pdf_history.js [hunk 16]
```diff
@@ -191,30 +192,31 @@
 
     _pushOrReplaceState: function pdfHistory_pushOrReplaceState(stateObj,
                                                                 replace) {
-//#if CHROME
-      // history.state.chromecomState is managed by chromecom.js.
-      if (window.history.state && 'chromecomState' in window.history.state) {
-        stateObj = stateObj || {};
-        stateObj.chromecomState = window.history.state.chromecomState;
+      if (typeof PDFJSDev === 'undefined' || PDFJSDev.test('CHROME')) {
```
Comment (Snuffleupagus) [inline web/pdf_history.js]: Nit: shouldn't this condition be `typeof PDFJSDev !== 'undefined' && PDFJSDev.test('CHROME')`, since it only matters in `CHROME` builds?

**Comment by Snuffleupagus** [comment]
Landing with r=bdahl on IRC.
Great effort to finally getting this done!
